Classify orchestrator child stderr and throttle skip noise

## Context

Operator observation on 2026-05-16 showed `npm run ai:orchestrator` running a large Knowledge Base sync and logging two separate noise classes:

- repeated `Skipping knowledge base sync; task already running (PID: 2082).` lines every poll while the existing sync child was still active
- normal KB embedding progress emitted through the daemon as `[ERROR] [ProcessSupervisor] knowledge base sync stderr: [LOG] Processed and embedded batch 83 of 237`

The embedding model/vector-dimension question is already closed separately: qwen3-8b @ 4096 dimensions is confirmed by healthcheck, direct endpoint smoke, and Chroma sample inspection. This ticket is only about daemon observability quality while that sync runs.

## The Problem

`ProcessSupervisorService` currently treats all child `stderr` data as daemon-level `ERROR`, even when the child is using stderr as its normal log channel. That is correct enough for MCP/harness transport visibility, but it is wrong at the daemon supervisor boundary: an informational child line should not become an operator-facing daemon error.

The repeated already-running skip line is also low-value noise during long-running child tasks. It confirms the singleton guard is working, but printing the same message every orchestrator poll obscures real progress and real failures.

## The Architectural Reality

- `ai/daemons/services/ProcessSupervisorService.mjs:195` detects a task already running and logs an INFO skip every invocation.
- `ai/daemons/services/ProcessSupervisorService.mjs:207` spawns child tasks with stderr piped.
- `ai/daemons/services/ProcessSupervisorService.mjs:209` forwards every stderr chunk as `writeLog('ERROR', ...)` regardless of child log prefix.
- `ai/daemons/TaskDefinitions.mjs:72` defines `kbSync` as a child process label `knowledge base sync`, so `buildScripts/ai/syncKnowledgeBase.mjs` progress lines flow through this supervisor path.
- `ai/daemons/Orchestrator.mjs:295` uses the orchestrator poll cadence, so already-running state is checked repeatedly while a long child task is active.

Existing adjacent tickets are not equivalent:

- #10576 / #10579 handled durable KB sync logging and are closed.
- #10088 is broader post-merge KB sync automation.
- #9919 / #9920 cover daemonization and re-queueing.
- #11075 is deferred config cleanup.

## The Fix

Add a narrow `ProcessSupervisorService` logging policy for child task output:

1. Classify child stderr lines by explicit child log prefix before forwarding to daemon logs. Examples: `[LOG]` and `[INFO]` should map to daemon INFO; `[WARN]` to WARN; `[ERROR]` or unclassified stderr can remain ERROR unless a stronger local convention exists.
2. Preserve stderr capture. The child process should continue piping stderr so MCP/server/script log channels remain observable.
3. Throttle or dedupe the already-running skip line per task/PID/reason window so a long KB sync does not emit identical skip messages every poll.
4. Keep failure semantics intact: non-zero exits, spawn failures, and real child error lines still surface as ERROR and still record failed task outcomes.
5. Add focused unit coverage in `test/playwright/unit/ai/daemons/services/ProcessSupervisorService.spec.mjs` for classification and skip dedupe/throttle.

## Contract Ledger Matrix

| Target Surface | Source of Authority | Proposed Behavior | Fallback | Docs | Evidence |
|---|---|---|---|---|---|
| `ProcessSupervisorService` child stderr forwarding | This ticket + operator 2026-05-16 log sample | Map explicit child log prefixes to daemon severity before writing supervisor logs | Unclassified stderr remains ERROR | JSDoc on helper methods if non-obvious | Unit tests for `[LOG]`, `[INFO]`, `[WARN]`, `[ERROR]`, and unclassified stderr |
| Already-running skip logging | This ticket + `ProcessSupervisorService.runTask()` current behavior | Emit the first skip for a task/PID/reason and suppress duplicate skip log spam while the same child remains active | Continue recording skipped outcomes for health visibility | Inline comment only if the state key is not obvious | Unit test repeated `runTask()` calls with same running state |
| Failure semantics | Existing `ProcessSupervisorService` child lifecycle behavior | Spawn failure, child error event, non-zero close, and explicit `[ERROR]` child lines remain ERROR | Existing task outcome recording stays authoritative | Existing method JSDoc | Existing tests plus new severity test |

## Acceptance Criteria

- [ ] Child stderr lines prefixed `[LOG]` or `[INFO]` are written as daemon INFO, not ERROR.
- [ ] Child stderr lines prefixed `[WARN]` are written as daemon WARN.
- [ ] Child stderr lines prefixed `[ERROR]` and unclassified stderr continue to be written as daemon ERROR.
- [ ] Already-running skip logs are deduped or throttled per task/PID/reason window while still recording a skipped outcome if that is required for health state.
- [ ] Existing spawn failure, child error event, and non-zero exit logging remain ERROR.
- [ ] Unit tests cover severity classification and repeated already-running skip behavior.
- [ ] No changes remove stderr usage from MCP servers or KB sync scripts; this is a daemon supervisor classification fix only.

## Out of Scope

- Changing the embedding provider, vector dimensions, or KB content sync semantics.
- Removing stderr from MCP servers or CLI scripts.
- Reworking the full logging framework or introducing a new structured logger.
- Changing orchestrator poll cadence or KB sync interval defaults.
- Implementing post-merge auto-sync; that belongs to existing automation tickets.

## Related

- #10576 — KB sync observability: tee MCP child logger output to `.neo-ai-data/logs/` for tail-able progress
- #10579 — KB Sync: Implement stderr observability (Shape A)
- #10088 — Automate post-merge knowledge-base sync trigger
- #9919 — Implement fs.watch Daemonization for Autonomous Orchestrator
- #9920 — Agent Error Recovery & Re-Queueing for Orchestrator
- #11075 — Migrate orchestrator magic numbers + retention constants to config

Origin Session ID: 6ec143cb-2e5b-4964-94d6-eb28cb25bde2

Handoff Retrieval Hint: query_raw_memories(query="orchestrator ProcessSupervisor knowledge base sync stderr LOG batch task already running PID skip noise child stderr severity classification")

--- a/ai/daemons/services/ProcessSupervisorService.mjs
+++ b/ai/daemons/services/ProcessSupervisorService.mjs
@@ -182,6 +182,77 @@ export class ProcessSupervisorService extends Base {
     }
 
     /**
+     * Clears the duplicate-running log guard for a task once it starts or exits.
+     * @param {String} taskName Task key.
+     * @returns {void}
+     */
+    clearRunningSkipLogState(taskName) {
+        if (!this.runningSkipLogKeys) {
+            return;
+        }
+
+        const prefix = `${taskName}:`;
+
+        for (const key of this.runningSkipLogKeys) {
+            if (key.startsWith(prefix)) {
+                this.runningSkipLogKeys.delete(key);
+            }
+        }
+    }
+
+    /**
+     * Determines whether an already-running skip should be logged.
+     * @param {String} taskName Task key.
+     * @param {String} reason Scheduling reason.
+     * @param {Number|null} pid Running child process ID.
+     * @returns {Boolean}
+     */
+    shouldLogRunningSkip(taskName, reason, pid) {
+        this.runningSkipLogKeys ??= new Set();
+
+        const key = `${taskName}:${pid ?? 'unknown'}:${reason}`;
+
+        if (this.runningSkipLogKeys.has(key)) {
+            return false;
+        }
+
+        this.runningSkipLogKeys.add(key);
+
+        return true;
+    }
+
+    /**
+     * Maps child-process stderr log prefixes to daemon log severities.
+     * @param {String} line Child stderr line.
+     * @returns {String}
+     */
+    getChildLogLevel(line) {
+        if (/^\[(LOG|INFO)\](?:\s|$)/.test(line)) {
+            return 'INFO';
+        }
+
+        if (/^\[WARN\](?:\s|$)/.test(line)) {
+            return 'WARN';
+        }
+
+        return 'ERROR';
+    }
+
+    /**
+     * Writes child stderr lines using their child-provided severity prefix.
+     * @param {Object} task Task definition.
+     * @param {Buffer|String} data Stderr chunk.
+     * @returns {void}
+     */
+    writeChildStderr(task, data) {
+        const lines = data.toString().split(/\r?\n/).map(line => line.trim()).filter(Boolean);
+
+        for (const line of lines) {
+            this.writeLog?.(this.getChildLogLevel(line), `[ProcessSupervisor] ${task.label} stderr: ${line}`);
+        }
+    }
+
+    /**
      * Starts a child task and wires completion status back into task state and HealthService.
      * @param {String} taskName Task key.
      * @param {String} reason Scheduling reason.
@@ -193,11 +264,14 @@ export class ProcessSupervisorService extends Base {
         const state = this.taskStateService.getTaskState(taskName);
 
         if (state.running) {
-            this.writeLog?.('INFO', `[ProcessSupervisor] Skipping ${task.label}; task already running (PID: ${state.pid}).`);
+            if (this.shouldLogRunningSkip(taskName, reason, state.pid)) {
+                this.writeLog?.('INFO', `[ProcessSupervisor] Skipping ${task.label}; task already running (PID: ${state.pid}).`);
+            }
             this.recordTaskOutcome(taskName, 'skipped', {reason, pid: state.pid, skippedAt: new Date().toISOString()});
             return false;
         }
 
+        this.clearRunningSkipLogState(taskName);
         this.taskStateService.markStarted(taskName, reason);
 
         this.writeLog?.('INFO', `[ProcessSupervisor] Starting ${task.label} (${reason}).`);
@@ -207,7 +281,7 @@ export class ProcessSupervisorService extends Base {
             child = this.spawnFn(task.command, task.args, {stdio: ['ignore', 'ignore', 'pipe']});
 
             child.stderr?.on('data', data => {
-                this.writeLog?.('ERROR', `[ProcessSupervisor] ${task.label} stderr: ${data.toString().trim()}`);
+                this.writeChildStderr(task, data);
             });
         } catch (e) {
             this.taskStateService.markSpawnFailed(taskName);
@@ -241,6 +315,7 @@ export class ProcessSupervisorService extends Base {
                     fs.unlinkSync(pidFile);
                 }
             } catch (e) {}
+            this.clearRunningSkipLogState(taskName);
 
             if (error) {
                 this.taskStateService.markFailed(taskName, null);

--- a/test/playwright/unit/ai/daemons/services/ProcessSupervisorService.spec.mjs
+++ b/test/playwright/unit/ai/daemons/services/ProcessSupervisorService.spec.mjs
@@ -8,6 +8,8 @@ import { ProcessSupervisorService } from '../../../../../../ai/daemons/services/
 function createTestService() {
     const dataDir = `/tmp/process-supervisor-service-test-${Date.now()}-${Math.random()}`;
     fs.ensureDirSync(dataDir);
+    const logEntries = [];
+    const taskOutcomes = [];
     
     const taskDefinitions = {
         mockTask: {
@@ -31,7 +33,7 @@ function createTestService() {
     };
 
     const mockHealthService = {
-        recordTaskOutcome: () => {}
+        recordTaskOutcome: (taskName, status, details) => taskOutcomes.push({details, status, taskName})
     };
 
     const service = Neo.create(ProcessSupervisorService, {
@@ -39,12 +41,12 @@ function createTestService() {
         taskDefinitions,
         taskStateService: mockTaskStateService,
         healthService: mockHealthService,
-        writeLog: () => {},
+        writeLog: (level, message) => logEntries.push({level, message}),
         spawnFn: () => ({ pid: 1234, on: () => {} }),
         processCommand: (pid) => 'echo hello'
     });
     
-    return { service, dataDir, mockTaskStateService };
+    return { service, dataDir, logEntries, mockTaskStateService, taskOutcomes };
 }
 
 test.describe('Neo.ai.daemons.services.ProcessSupervisorService', () => {
@@ -94,5 +96,52 @@ test.describe('Neo.ai.daemons.services.ProcessSupervisorService', () => {
         
         expect(result).toBe(false);
         expect(spawnCalled).toBe(false);
+    });
+
+    test('runTask classifies child stderr log prefixes', () => {
+        const { service, logEntries } = createTestService();
+        let stderrHandler;
+
+        service.spawnFn = () => ({
+            pid   : 9999,
+            stderr: {
+                on: (eventName, handler) => {
+                    if (eventName === 'data') {
+                        stderrHandler = handler;
+                    }
+                }
+            },
+            on: () => {}
+        });
+
+        service.runTask('mockTask', 'test-reason');
+
+        stderrHandler(Buffer.from([
+            '[LOG] Processed and embedded batch 83 of 237',
+            '[INFO] Sync still running',
+            '[WARN] Slow embedding batch',
+            '[ERROR] Failed embedding batch',
+            'plain stderr output'
+        ].join('\n')));
+
+        const stderrLogs = logEntries.filter(entry => entry.message.includes('stderr:'));
+
+        expect(stderrLogs.map(entry => entry.level)).toEqual(['INFO', 'INFO', 'WARN', 'ERROR', 'ERROR']);
+    });
+
+    test('runTask dedupes repeated already-running skip logs', () => {
+        const { service, logEntries, taskOutcomes } = createTestService();
+
+        service.taskStateService.getTaskState = () => ({ running: true, pid: 1234 });
+
+        service.runTask('mockTask', 'same-reason');
+        service.runTask('mockTask', 'same-reason');
+        service.runTask('mockTask', 'different-reason');
+
+        const skipLogs = logEntries.filter(entry => entry.message.includes('task already running'));
+        const skippedOutcomes = taskOutcomes.filter(entry => entry.status === 'skipped');
+
+        expect(skipLogs.length).toBe(2);
+        expect(skippedOutcomes.length).toBe(3);
     });
 });

